### ISSUE TITLE
[Feat] 테마 지역 외 여행지 출력 제거

### ISSUE THREAD

**Issue by thisischeese** [manru-homerun/AI#31]
<html>
<body>
<html><head></head><body><h2>🥕 작업 배경</h2><ul><li><p>현재 GRU 모델은 사용자 특성과 기존 여행지 Sequence를 기반으로 전체 POI Vocabulary에 대해 다음 여행지의 Score를 예측한다.</p></li><li><p>학습 데이터는 실제 방문 로그 전체를 그대로 사용하지 않고, <em></em>서비스에서 지원 가능한 POI만 추출한 Ordered Subsequence를 사용해 구성하였다.</p><pre><code class="language-text">실제 방문 로그
A → X → B → Y → C

서비스 지원 POI
A, B, C

학습 Sequence
A → B → C
</code></pre></li><li><p>따라서 모델이 학습하는 Sequence는 실제 바로 다음 방문지를 의미하기보다, <em></em>서비스에서 지원 가능한 POI 중 다음으로 관측되는 여행지의 순서를 의미한다.</p></li><li><p>동일한 기준을 추론에도 적용하기 위해 <code inline="">/recommend</code> 요청의 <code inline="">contentIdSequence</code>에서도 <strong>Model Vocabulary에 존재하는 POI만 순서를 유지한 채 추출하여 GRU Prefix로 사용</strong>한다.</p></li><li><p>요청 Sequence에 미학습 POI가 포함되어 있더라도 전체 요청을 Fallback 처리하지 않고, 학습 전처리와 동일하게 해당 POI를 제외한 Known Sequence를 사용한다.</p></li><li><p>단, Model Vocabulary에 포함되는 POI가 하나도 없어 **Known Sequence가 비어 있는 경우에는 GRU가 활용할 Sequence 정보가 없으므로 기존 지역 중심 Top-4 추천으로 Fallback<em></em>한다.</p></li><li><p>또한 현재 GRU는 전체 POI Vocabulary를 대상으로 다음 여행지 Score를 계산하기 때문에 요청된 <code inline="">areaCode</code>와 무관한 POI도 후보군에 포함될 수 있다.</p></li><li><p>지역 제한과 유효 POI 여부는 모델이 학습을 통해 스스로 지키도록 하기보다, <em></em>추론 단계의 Hard Constraint로 처리하여 서비스 요구사항을 보장**한다.</p></li></ul><hr><h2>🥕 작업 목표</h2><ul><li><p><code inline="">/recommend</code>의 <code inline="">contentIdSequence</code>에서 <strong>Model Vocabulary에 존재하는 POI만 순서를 유지하여 Known Sequence를 생성</strong>한다.</p></li><li><p>Known Sequence가 존재하는 경우 해당 Sequence를 GRU Prefix로 사용해 다음 여행지를 추천한다.</p></li><li><p>Known Sequence가 비어 있는 경우 기존 <strong>기초지자체 중심 관광지 Top-4 추천 로직으로 Fallback</strong>한다.</p></li><li><p><code inline="">areaCode</code>에 해당하는 POI만 모델의 최종 후보군으로 선택할 수 있도록 <strong>지역 기반 Logit Masking</strong>을 적용한다.</p></li><li><p>모델이 학습한 전체 POI 중 <strong>POI Master에 등록된 유효한 <code inline="">contentId</code>만 추천 및 생성 가능하도록 제한</strong>한다.</p></li><li><p><code inline="">/recommend</code>, <code inline="">/generate-course</code>의 모델 출력에는 동일한 Candidate Constraint를 적용하여 다음 조건을 보장한다.</p><ul><li><p>테마 지역 외 POI 출력 방지</p></li><li><p>POI Master에 존재하지 않는 <code inline="">contentId</code> 출력 방지</p></li><li><p>이미 여행계획에 포함된 POI 중복 추천 방지</p></li></ul></li></ul><hr><h2>🥕 작업 내용</h2><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" checked="" disabled=""> <strong>POI Master 구축</strong></p><ul><li><p>학습 데이터에 존재하는 <code inline="">content_id</code>를 기준으로 POI Master 생성</p></li><li><p>최소 관리 컬럼</p><ul><li><p><code inline="">content_id</code></p></li><li><p><code inline="">token_id</code></p></li><li><p><code inline="">area_code</code></p></li><li><p><code inline="">area_name</code></p></li><li><p><code inline="">content_name</code></p></li></ul></li><li><p><code inline="">final_travel_sequence.csv</code>에서 지역 정보가 누락된 POI는 한국관광공사 API 등을 통해 <code inline="">area_code</code> 보완</p></li><li><p>지역을 확인할 수 없는 POI는 모델 후보군에서 제외</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>Model Vocabulary Mapping 관리</strong></p><ul><li><p>학습 시 사용한 <code inline="">content_id ↔ token_id</code> Mapping을 추론에서도 동일하게 사용</p></li><li><p>예:</p><pre><code class="language-text">content_to_token.json
token_to_content.json
</code></pre></li><li><p>POI Master와 Model Vocabulary의 역할을 구분</p><ul><li><p>POI Master: 모델이 출력 가능한 POI 및 지역 정보 관리</p></li><li><p>Model Vocabulary: GRU가 Embedding과 Token을 가지고 있는 POI 관리</p></li></ul></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>지역 코드 정규화</strong></p><ul><li><p>API의 <code inline="">areaCode</code>와 POI Master의 지역 코드가 동일한 기준으로 비교되도록 변환 로직 구현</p></li><li><p>예:</p><pre><code class="language-text">11000 → 서울
26000 → 부산
</code></pre></li><li><p>서비스 지역 코드와 학습 데이터의 지역명 Mapping 관리</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong><code inline="">contentIdSequence</code> → Known Sequence 변환</strong></p><ul><li><p><code inline="">/recommend</code> 요청의 <code inline="">contentIdSequence</code>에서 Model Vocabulary에 존재하는 POI만 추출</p></li><li><p>기존 여행 순서는 그대로 유지</p></li><li><p>미학습 POI는 <code inline="">UNK</code> Token으로 변환하지 않고 Sequence에서 제외</p></li></ul><pre><code class="language-python">known_sequence = [
    content_id
    for content_id in request.contentIdSequence
    if content_id in content_to_token
]
</code></pre><ul><li><p>예:</p></li></ul><pre><code class="language-text">요청 Sequence
A → X → B → Y → C

A, B, C : Model Vocabulary 존재
X, Y    : Model Vocabulary 미존재

        ↓

Known Sequence
A → B → C
</code></pre><ul><li><p>이는 학습 데이터 생성 시 적용한 <strong>서비스 지원 POI Ordered Subsequence 추출 방식과 동일한 정책</strong>으로 처리한다.</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>Known Sequence 기반 <code inline="">/recommend</code> 분기 처리</strong></p><ul><li><p><code inline="">contentIdSequence</code>가 비어 있거나 전체 여행 방문 가능 개수에 도달한 경우 기존 지역 Top-4 로직 사용</p></li><li><p><code inline="">0 &lt; len(contentIdSequence) &lt; travelDuration × 6</code>인 경우 Known Sequence 생성</p></li><li><p>Known Sequence가 하나 이상 존재하면 GRU 추론 수행</p></li><li><p>Known Sequence가 비어 있으면 기초지자체 중심 관광지 Top-4 Fallback 수행</p></li></ul><pre><code class="language-text">/recommend
     │
     ▼
contentIdSequence 길이 확인
     │
     ├─ 0
     │    └─→ 지역 중심 Top-4
     │
     ├─ travelDuration × 6 이상
     │    └─→ 지역 중심 Top-4
     │
     └─ 그 외
          │
          ▼
   Model Vocabulary 기준
   Known Sequence 추출
          │
     ┌────┴────┐
     │         │
   존재       없음
     │         │
     ▼         ▼
    GRU      지역 중심
            Top-4 Fallback
</code></pre></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong><code inline="">/recommend</code> GRU 추론 로직 수정</strong></p><ul><li><p>Known Sequence를 Token Sequence로 변환하여 GRU Prefix로 입력</p></li><li><p>GRU의 다음 POI Logit 계산</p></li><li><p><code inline="">지역 Mask → 유효 POI Mask → 중복 Mask</code> 순으로 적용</p></li><li><p>최종 후보 중 Score가 높은 POI Top 4 반환</p></li></ul><pre><code class="language-text">contentIdSequence
A → X → B → Y

     ↓ Model Vocabulary Filtering

Known Sequence
A → B

     ↓

    GRU

     ↓

Area / Valid / Duplicate Mask

     ↓

   Top-4
</code></pre></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>지역별 POI Candidate Mask 생성</strong></p><ul><li><p>POI Vocabulary 전체 크기와 동일한 Boolean Mask 생성</p></li><li><p>요청 <code inline="">areaCode</code>에 속한 POI만 <code inline="">True</code>로 설정</p></li><li><p>해당 지역 외 POI의 Logit을 <code inline="">-inf</code>로 변경</p></li></ul><pre><code class="language-python">logits[~area_mask] = float("-inf")
</code></pre></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>유효 POI Mask 적용</strong></p><ul><li><p>POI Master에 존재하지 않는 Token은 항상 후보에서 제외</p></li><li><p>GRU가 임의의 <code inline="">contentId</code>를 생성하는 것이 아니라 POI Vocabulary의 등록된 Token 중에서만 선택하도록 제한</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>기존 여행지 중복 Mask 적용</strong></p><ul><li><p><code inline="">/recommend</code>에서는 <strong>Known Sequence뿐 아니라 원본 <code inline="">contentIdSequence</code> 전체를 제외 대상</strong>으로 사용</p></li><li><p><code inline="">/generate-course</code>에서는 이미 생성된 여행지 및 <code inline="">contentIdList</code>를 제외</p></li><li><p>Model Vocabulary에 존재하는 해당 POI의 Logit을 <code inline="">-inf</code> 처리하여 동일 여행지가 다시 추천되지 않도록 구현</p></li></ul><pre><code class="language-text">입력 Sequence
A → X → B

GRU Prefix
A → B

추천 제외 대상
A, X, B
</code></pre><ul><li><p>단, <code inline="">X</code>가 Model Vocabulary에 존재하지 않으면 애초에 모델 출력 후보가 될 수 없으므로 별도 Token Mask는 불필요</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong><code inline="">/generate-course</code> 추론 로직 수정</strong></p><ul><li><p><code inline="">travelDuration × 6</code>을 최종 생성 개수로 설정</p></li><li><p>GRU로 다음 POI를 하나씩 Autoregressive하게 생성</p></li><li><p>매 Step마다 지역/유효 POI/중복 Mask를 다시 적용</p></li><li><p>사용자가 전달한 <code inline="">contentIdList</code>는 반드시 최종 여행계획에 포함되도록 별도 Required POI로 관리</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong><code inline="">contentIdList</code> 사전 검증</strong></p><ul><li><p>요청된 필수 관광지가 POI Master에 존재하는지 확인</p></li><li><p>요청 <code inline="">areaCode</code>와 동일한 지역에 속하는지 확인</p></li><li><p>존재하지 않거나 다른 지역의 <code inline="">contentId</code>가 포함된 경우 모델 호출 전에 예외 처리</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>Masking 공통 모듈화</strong></p><ul><li><p><code inline="">/recommend</code>, <code inline="">/generate-course</code>에서 동일한 Candidate Constraint를 사용할 수 있도록 공통 함수로 분리</p></li></ul><pre><code class="language-python">apply_candidate_mask(
    logits,
    area_code,
    excluded_content_ids,
)
</code></pre></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>Fallback 처리</strong></p><ul><li><p><code inline="">/recommend</code>에서 Known Sequence가 비어 있는 경우 기초지자체 중심 관광지 Top-4 Fallback 수행</p></li><li><p>Mask 적용 후 추천 가능한 후보가 부족한 경우에도 기존 지역 기반 Fallback을 통해 보완</p></li><li><p>후보 부족 상황에서도 다른 지역 POI를 반환하는 방식으로 Constraint를 완화하지 않음</p></li></ul></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <strong>테스트 코드 작성</strong></p><ul><li><p><code inline="">contentIdSequence</code>가 전부 Model Vocabulary에 존재하는 경우 전체 Sequence가 GRU에 입력되는지 검증</p></li><li><p>일부 POI만 Model Vocabulary에 존재하는 경우 순서를 유지한 Known Sequence가 생성되는지 검증</p></li><li><p>모든 POI가 Model Vocabulary에 없는 경우 Top-4 Fallback이 호출되는지 검증</p></li><li><p>미학습 POI가 <code inline="">UNK</code> Token으로 변환되지 않는지 검증</p></li><li><p>다른 지역 POI 출력 여부 검증</p></li><li><p>POI Master에 없는 <code inline="">contentId</code> 출력 여부 검증</p></li><li><p>동일 POI 중복 추천 여부 검증</p></li><li><p><code inline="">contentIdList</code>가 최종 생성 결과에 모두 포함되는지 검증</p></li><li><p>Mask 이후 후보가 부족한 경우 Fallback 정상 동작 여부 검증</p></li></ul></li></ul><hr><h2>🥕 <code inline="">/recommend</code> 최종 정책</h2>
contentIdSequence 상태 | 처리
-- | --
비어 있음 | 지역 중심 관광지 Top-4
전체 방문 가능 개수 이상 | 지역 중심 관광지 Top-4
일부/전체 POI가 Model Vocabulary에 존재 | Known Sequence 추출 후 GRU
Model Vocabulary에 존재하는 POI가 0개 | 지역 중심 관광지 Top-4

<p>예:</p><pre><code class="language-text">[A, B, C]
모두 Known
→ A → B → C
→ GRU
</code></pre><pre><code class="language-text">[A, X, B, Y]
A, B만 Known
→ A → B
→ GRU
</code></pre><pre><code class="language-text">[X, Y, Z]
모두 Unknown
→ Known Sequence 없음
→ 지역 중심 Top-4 Fallback
</code></pre><hr><h2>🥕 완료 조건</h2><ul><li><p><code inline="">contentIdSequence</code>에서 Model Vocabulary에 존재하는 POI만 <strong>순서를 유지한 Ordered Subsequence</strong>로 GRU에 입력</p></li><li><p>Known Sequence가 비어 있을 경우 GRU 호출률 <strong>0%</strong></p></li><li><p>Known Sequence가 비어 있을 경우 지역 중심 Top-4 Fallback 정상 수행률 <strong>100%</strong></p></li><li><p>미학습 POI의 <code inline="">UNK</code> Token 변환 <strong>0건</strong></p></li><li><p><code inline="">areaCode</code>와 다른 지역의 POI 출력률 <strong>0%</strong></p></li><li><p>POI Master에 존재하지 않는 <code inline="">contentId</code> 출력률 <strong>0%</strong></p></li><li><p>하나의 여행계획 내 동일 POI 중복률 <strong>0%</strong></p></li><li><p>유효한 <code inline="">contentIdList</code>의 최종 여행계획 포함률 <strong>100%</strong></p></li><li><p><code inline="">/recommend</code>, <code inline="">/generate-course</code>의 모델 출력에 동일한 POI Candidate Constraint 적용</p></li></ul></body></html>
</body>
</html>
